email/fetcher.go

```diff
@@ -0,0 +1,393 @@
+		// we reply to the original email those replies are picked up by the
+		// scanner as well
+		if isFromAbuseScanner(msg) {
+			logger.Debugf("skip message from abuse scanner (expected)")
+			err := f.persistSkipMessage(mailbox, msg)
+			if err != nil {
+				logger.Errorf("Failed to persist skip message, error: %v\n", err)
+			}
+			continue
+		}
+
+		// skip messages without body
+		//
+		// TODO: side-effect from UidFetch and can probably be avoided
+		if !hasBody(msg) {
+			logger.Debugf("skip message due to not having a body (expected)")
+			err := f.persistSkipMessage(mailbox, msg)
+			if err != nil {
+				logger.Errorf("Failed to persist skip message, error: %v\n", err)
+			}
+			continue
+		}
+
+		toUnsee.AddNum(msg.Uid)
+		err := f.persistMessage(mailbox, msg, section)
+		if err != nil {
+			logger.Errorf("Failed to persist %v, error: %v\n", msg.Uid, err)
+		}
+	}
+
+	// unsee messages
+	flags := []interface{}{imap.SeenFlag}
+	err = email.UidStore(toUnsee, "-FLAGS.SILENT", flags, nil)
+	if err != nil && !strings.Contains(err.Error(), "Could not parse command") {
+		logger.Debugf("Failed to unsee messages, error: %v\n", err)
+	} else {
+		logger.Debugln("Successfully unseen messages")
+	}
+
+	return <-done
+}
+
+// persistMessage will persist the given message in the abuse scanner database
+func (f *Fetcher) persistMessage(mailbox *imap.MailboxStatus, msg *imap.Message, section *imap.BodySectionName) error {
+	// convenience variables
+	abuseDB := f.staticDatabase
+
+	// build the uid
+	uid := buildMessageUID(mailbox, msg.Uid)
+
+	// read the entire message body
+	bodyLit := msg.GetBody(section)
+	if bodyLit == nil {
+		return fmt.Errorf("msg %v has no body", uid)
+	}
+	body, err := ioutil.ReadAll(bodyLit)
+	if err != nil {
+		return errors.AddContext(err, "could not read msg body")
+	}
+
+	// parse the 'from'
+	from := "unknown"
+	if len(msg.Envelope.From) > 0 {
+		from = msg.Envelope.From[0].Address()
+	}
+
+	// create the email entity from the message
+	email := database.AbuseEmail{
+		ID:        primitive.NewObjectID(),
+		UID:       uid,
+		UIDRaw:    msg.Uid,
+		Body:      body,
+		From:      from,
+		Subject:   msg.Envelope.Subject,
+		MessageID: msg.Envelope.MessageId,
+
+		Parsed:    false,
+		Blocked:   false,
+		Finalized: false,
+
+		InsertedAt: time.Now(),
+	}
+
+	// insert the message in the database
+	err = abuseDB.InsertOne(email)
+	if err != nil {
+		return errors.AddContext(err, "could not insert email")
+	}
+	return nil
+}
+
+// persistSkipMessage will persist the given message as finalized in the abuse
+// scanner database, this ensures the message won't be considered 'missing'
+func (f *Fetcher) persistSkipMessage(mailbox *imap.MailboxStatus, msg *imap.Message) error {
+	// convenience variables
+	abuseDB := f.staticDatabase
+
+	// build the uid
+	uid := buildMessageUID(mailbox, msg.Uid)
+
+	// skip if it already exists
+	existing, err := abuseDB.FindOne(uid)
```

Comment:
This is not an effective way to ensure the email is not there. This service is likely to work on multiple machines, so race conditions are possible.

A more resilient approach would be to make UID a unique field, try to insert, and then check the error on insertion for "record exists".

Reply:
Wanted to do that in a F/U but I tackled it here. I implemented it a bit differently. I don't want to do the insert as it would leave a record that I would have to then update. I just used the locks.